README.md
##Basestar
A javascript in-browser development environment powered by [nodejs](http://nodejs.org/).

##Features:

- Code editor with tabbed window support
- Project browser
- Code navigator
- History for in-browser and server-side
- Create new project from git repository
- Control script (run/stop) in the ide
- Console window for output
- Simple profiling info
- Colorful diff
- Http monitor
- Git operations (commit, push, pull etc.)
- Debugging (breakpoint, inspector)
- Env.js
- Vim key-bindings (emac?)
- Test suite creator and runner

##Technology

- [UKI](http://github.com/voloko/uki) (Plan to change)
- [Codemirror](http://codemirror.net/)
- [SocketIO](http://socket.io/)

## How To Contribute
If you have an idea which will **dramatically** change the direction of the project or the effect is long-term, you could `fork` the project,
`put` the idea in this readme and `send` out a pull request. (Why? Checkout [RDD](http://tom.preston-werner.com/2010/08/23/readme-driven-development.html))

If you have more specific suggestion or features want to implement,
then you can put them in file **CHANGELOG.md** under an existent version or create a new version if you think make sense,
and follow the same step like above (fork, change, send pull request). Here we may follow the concept: `Changelog for future = todo list`.

##History

This project was started for [nodeknockout](http://nodeknockout.com/) - a nodejs coding contest.
You can go to the [team page](http://nodeknockout.com/teams/fewtter) for more about the project history.
After the contest, [Eric](https://github.com/ericfong) gave this new name **Basestar** - [a starship in Battlestar Galactica](http://en.wikipedia.org/wiki/Cylon_Basestar).

## License

(The MIT License)

Copyright (c) 2010 Senmiao Liu <zir.echo@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

CHANGELOG.md
0.0.1 (?)
---------
- Update README.md for new project description.


app.js
// Main application module


// Dependencies and settings
var Path = require("path"),
fs = require("fs"),
FileHandler = genji.web.handler.FileHandler,
staticRoot = Path.join(__dirname, "/static"),
viewRoot = Path.join(__dirname, "/view"),
nun = require("nun"),
chain = genji.pattern.control.chain,
workspace = genji.settings.workspace,
parseQuery = require("querystring").parse,
EventEmitter = require("events").EventEmitter
emitter = new EventEmitter;

/* http endpoint */

function index(handler) {
    var ctx = {
        staticUrl: genji.settings.staticUrl
    }
    _render("/index.xhtml", ctx, function(html) {
        handler.sendHTML(html);
    })
}

function filetree(handler) {
    if (genji.emitter)
        genji.emitter.emit('message', 'file tree called');
    handler.on('end', function(data) {
        var dir = parseQuery(data).dir;
        if (dir) {
            var fullDir = Path.join(workspace, dir);
            fs.stat(fullDir, function(err, stats) {
                if (err) {
                    handler.error(500, err.toString());
                    return;
                }
                if (stats.isDirectory()) {
                    var results = ['<ul class="jqueryFileTree" style="display: none;">'];
                    fs.readdir(fullDir, function(err, files) {
                        if (err) {
                            handler.error(500, err.toString());
                            return;
                        }
                        if (files.length == 0) {
                            results.push('</ul>');
                            handler.send(results.join(''));
                            return;
                        }
                        chain(files, function(file, idx, arr, next) {
                            var f = Path.join(fullDir, file);
                            fs.stat(f, function(err, stats) {
                                if (err) {
                                    handler.error(500, err.toString());
                                    return;
                                }
                                if (file[0] != '.') {
                                    var relativePath = Path.join(dir, file);
                                    if (stats.isDirectory()) {
                                        results.push('<li class="directory collapsed"><a href="#" rel="'+relativePath+'/">'+file+'</a></li>');
                                    }
                                    if (stats.isFile()) {
                                        var ext = Path.extname(file).slice(1);
                                        results.push('<li class="file ext_'+ext+'"><a href="#" rel="'+relativePath+'">'+file+'</a></li>');
                                    }
                                }
                               next();
                               if (idx == files.length) {
                                   results.push('</ul>');
                                   handler.send(results.join(""));
                               }
                            });
                        })();
                    });
                } else {
                    handler.error(500, params.dir + " not a directory");
                }
            })
        } else {
            handler.error(404, "");
        }
    });
}

// serve static file under ./static
function staticFile(handler, filepath) {
    handler.setRoot(staticRoot);
    handler.staticFile(filepath);
}

/* Restful json api */

function gitClone(handler) {
    handler.on('end', function (data) {
        var params = parseQuery(data);
        if (!params.url || !params.name) {
            handler.send("Your must provide both url and name. Your input:\n" + JSON.stringify(params));
        } else {
            handler.sendHeaders(200, {'Content-Type': 'text/plain'});
            _gitClone(params.url, params.name, function(stdout, stderr) {
                if (stdout || stderr) {
                    handler.response.write((stdout || stderr) + "<br />");
                } else {
                    handler.finish();
                }
            });
        }
    });
}

function message(handler) {
    handler.sendHeaders(200, {'Content-Type': 'text/plain'});
    setTimeout(function() {
        handler.finish(new Date + '');
    }, 1000);
}

function getCode(handler, filename) {
    if (filename) {
        fs.readFile(Path.join(workspace, filename), 'utf8', function(err, data) {
            if (err) throw err;
            handler.sendJSON({code: data.length > 0 ? data : ''});
        });
    }
}

/* private functions */

// render mustache template by nun
function _render(path, context, callback) {
    nun.render(Path.join(viewRoot, path), context, {compress: false, cache: true}, function(err, result) {
       if (err) {
            throw err;
        }
        var buffer = '';
        result.addListener('data', function (c) {
            buffer += c;
        }).addListener('end', function () {
            callback(buffer);
        });
    });
}

var exec  = require('child_process').exec,
spawn = require('child_process').spawn,
nodeExec = genji.settings.node,
gitExec = genji.settings.git;

function _exec(cmd, options, callback) {
    var exec = spawn(cmd, options);
    exec.stdout.on('data', function (data) {
        callback('' + data);
    });

    exec.stderr.on('data', function (data) {
        callback(null, '' + data);
    });

    exec.on('exit', function (code) {
        callback(null, null, code);
        console.log(cmd + ': child process exited with code ' + code);
    });
    return exec;
}

function _gitCommand(options, callback) {
    _exec(gitExec, options, callback);
}

function _gitClone(url, name, callback) {
   var dir = Path.join(workspace, name);
   _gitCommand(["clone", "--recursive", url, dir], callback);
}

function _saveFile(client, data) {
    var path = Path.join(workspace, data.filename);
    fs.writeFile(path, data.content, 'utf8', function(err) {
        if (err) {
            client.send({type:'msg', data:['Error, cannot save file: ', data.filename]});
        } else {
            client.send({type:'msg', data:[data.filename + ' saved!']});
            client.pub({type:'msg', data:['Someone saved file ' + data.filename]});
        }
    });
}

function _node(client, data) {
    if (!data.filename) {
        client.send({type:'msg', data: ["Your must provide the script name, try open a file first."]});
        return;
    }
    client.pub({type: 'msg', data: ['Someone just run script: ' + data.filename]});
    client.send({type: 'msg', data: ["Your process can running for at most 120 seconds."]});
    var path = Path.join(workspace, data.filename);
    var options = data.options || [], id;
    options.push(path);
    var child = _exec(nodeExec, options, function(stdout, stderr, code) {
       var msg = stdout || stderr;
       msg && client.send({type: 'msg', data: [msg]});
       if (code !== undefined) {
          client.send({type: 'msg', data: ["script exited with: " + code]});
          clearTimeout(id);
       }
   });
   id = setTimeout(function() {
       child.kill('SIGKILL');
   }, 120000);
   
}

emitter.on("message", function(client, msg) {
    this.emit(msg.type, client, msg.data);
});

emitter.on("cmd", function(client, data) {
    if (data.cmd == 'gitClone') {
        if (data.url && data.name) {
            _gitClone(data.url, data.name, function(stdout, stderr, code) {
                var msg = stdout || stderr;
                msg && client.send({type: 'msg', data: [msg]});
                if (code === 0) {//cloned
                    // tell client to refresh project browser
                    var m = ["User " + client.sessionId, "cloned", data.url,"as project",data.name];
                    client.pub({type: 'msg', data:[m.join(" ")]})
                    // for others
                    client.pub({type:'event', name: 'newProjectCreated', data:[]});
                    // for youself
                    client.send({type:'event', name: 'newProjectCreated', data:[]});
                }
            });
        } else {
            client.send({type:'msg', data: ["Your must provide both url and name."]})
        }
    }
    if (data.cmd == 'saveFile') {
        _saveFile(client, data);
    }
    if (data.cmd == 'node') {
        _node(client, data);
    }
});

// message hub
process.nextTick(function() {
if ('io' in genji) {
    var buffer = [];
    genji.io.on('connection', function(client){
        // record history for broadcast
        var broadcast = function(msg) {
            if (msg.type == 'msg') {
                msg.data.push(new Date+'');
                buffer.push(msg);
            }
            if (buffer.length > 30) buffer.shift();
            client.broadcast(msg);
        }
        client.pub = broadcast;
        if (buffer.length > 0) {
            buffer.forEach(function(msg) {
                client.send(msg);
            })
        }
        broadcast({type: 'msg', data: [client.sessionId + ' connected']});
        client.on('message', function(msg) {
            emitter.emit("message", client, msg);
        });

        client.on('disconnect', function(){
            broadcast({type: 'msg', data: [client.sessionId + ' disconnected']});
        });
    });
}
});


module.exports = [
    ["^/$", index, "get"],
    ["^/api/git/clone/$", gitClone , "post"],
    ["^/api/filetree/$", filetree, "post"],
    ["^/api/message/$", message, "get"],
    ["^/api/getCode/-(.*)$", getCode, "get"],
    [FileHandler, "^/static/(.*)$", staticFile, "get"]
];

server.js
var Path = require('path');

// basic settings
var settings = {
    libPath: ["/lib/genji/lib", "/lib", "/lib/node-git/lib", "/lib/socket.io-node/lib/"]
    ,staticUrl: 'http://fewtter.no.de/static/'
    ,env: {type: 'production', root: __dirname, level: 1}
    ,servers: [
        {host: '8.19.40.199', port: 80}
    ]
    ,git: '/opt/local/bin/git'
    ,node: '/usr/nodejs/latest/bin/node'
    ,workspace: "/home/node/workspace"
    ,middlewares: [
        {name:'response-time'},
        {name: 'error-handler'},
        {name:'logger'},
        {name:'conditional-get'},
        {name: 'router', conf: {handler: 'genji.web.handler.SimpleCookieHandler', urls: './app'}}
    ]
};

// try to merge with local settings defined in ./config.js
try {
    var local_settings = require('./config');
    for (var key in local_settings) {
        settings[key] = local_settings[key];
    }
} catch(e) {
    require('sys').puts(e);
}

// push dependencies into require path
settings.libPath.forEach(function(path) {
    require.paths.unshift(Path.join(__dirname, path));
});

// load `genji` and start server
var genji = require("genji");
var p = genji.util.manager.startServer(settings);

if (p.server) {
    var socket = require('socket.io');
    var io = socket.listen(p.server);
    genji.io = io;
}

static/js/ft.js
;
(function($) {
    $.ft.evt = $({});

    $.ft.buildFileTree = function() {
        $('#panelProject').fileTree({
            root: '/',
            script: '/api/filetree/'
        }, function(file) {
            $.ajax({
                url: '/api/getCode/-' + file,
                type: 'get',
                success: function (data) {
                    $.ft.createEditor(file, data.code);
                }
            });
        });
    }

    $.ft.init = function() {
        // build the main layout
        $.ft.buildLayout();
        // build the file browser
        $.ft.buildFileTree();
        // save editors in an object
        $.ft.editors = {};
        /* action for buttons */
        $('#btnNewProject').click(function() {
            // toggle the `new project` panel
            uki('#ppNewProject')[0].toggle();
        });
        // tell server to clone the project
        uki('#btnGitClone').click(function() {
            $.ft.socket.send({type: 'cmd', data: {
                    cmd: 'gitClone',
                    url: uki('#inputProjectUrl').value(),
                    name: uki('#inputProjectName').value()
            }});
        });
        $('#btnVote').append('<a href="http://nodeknockout.com/teams/fewtter" target="nko" title="Help me win Node.js KO!"><img style="position: fixed; border: 0px;" src="http://nodeknockout.com/images/voteko.png" alt="Help me win Node.js KO!" /></a>');

        // save file to server
        uki('#btnSave').click(function() {
            var filename = $.ft.currentFile;
            if (filename) {
                $.ft.socket.send({type:'cmd',
                    data: {cmd: 'saveFile', filename: $.ft.currentFile, content: $.ft.editors[filename].editor.getCode()}});
            } else {
                $.ft.log("Please open a file first");
            }
        });

        // run script with node
        uki('#btnRun').click(function() {
            var filename = $.ft.currentFile;
            if (filename) {
                    $.ft.socket.send({type:'cmd',
                        data: {cmd: 'node', filename: $.ft.currentFile}});
            } else {
                $.ft.log("Please open a file first");
            }
        });

        // setup socket.io
        io.setPath('/static/js/socket.io/');
        var socket = new io.Socket(null, {
            transports: ['websocket', 'htmlfile', 'xhr-multipart', 'xhr-polling']
        });
        socket.connect();
        socket.on('message', function(msg) {
            if (msg.type == 'msg') {
               msg.data.forEach(function(m) {
                   $.ft.log(m);
               })
            }
            if (msg.type == 'event') {
               $.ft.evt.trigger(msg.name, msg.data);
            }
        });
        $.ft.socket = socket;

        // handle events emitted from server
        $.ft.evt.bind('newProjectCreated', function() {
            // refresh the file browser
            $.ft.buildFileTree();
        });
    
        // handle global ajax error
        
        $.ajax({
            ajaxError: function(xhr, status) {
                $.ft.log("error: " + status);
            }
        });
    }

    $.ft.createEditor = function(filename, code) {
        var editors = $.ft.editors;
        $.ft.currentFile = filename;
        uki('#codeEditorLabel').text('Editor ' + filename);
        if (filename in editors) {
            editors[filename].editor.setCode(code);
        } else {
            var wrapper = $.ft.createEditorWrapper(filename);
           //  create code editor
            wrapper.attachTo(document.getElementById('codeEditor'));
            var editor = new CodeMirror(wrapper.dom(), {
                height: "98%",
                width: '99%',
                content: code,
                parserfile: ["tokenizejavascript.js", "parsejavascript.js"],
                stylesheet: ["static/js/codemirror/css/jscolors.css"],
                path: "static/js/codemirror/js/",
                autoMatchParens: true
            });
            editors[filename] = wrapper.dom();
            editors[filename].editor = editor;
        }
        // show current and hide others
        for (var e in editors) {
            e == filename ? $(editors[e]).css('visibility', 'visible') : $(editors[e]).css('visibility', 'hidden');
        }
    }

    $.ft.log = function(data) {
        $('#panelConsole').prepend('<p>' + data + '</p>');
    }

})(jQuery);

static/js/jqueryFileTree.js
// jQuery File Tree Plugin
//
// Version 1.01
//
// Cory S.N. LaViska
// A Beautiful Site (http://abeautifulsite.net/)
// 24 March 2008
//
// Visit http://abeautifulsite.net/notebook.php?article=58 for more information
//
// Usage: $('.fileTreeDemo').fileTree( options, callback )
//
// Options:  root           - root folder to display; default = /
//           script         - location of the serverside AJAX file to use; default = jqueryFileTree.php
//           folderEvent    - event to trigger expand/collapse; default = click
//           expandSpeed    - default = 500 (ms); use -1 for no animation
//           collapseSpeed  - default = 500 (ms); use -1 for no animation
//           expandEasing   - easing function to use on expand (optional)
//           collapseEasing - easing function to use on collapse (optional)
//           multiFolder    - whether or not to limit the browser to one subfolder at a time
//           loadMessage    - Message to display while initial tree loads (can be HTML)
//
// History:
//
// 1.01 - updated to work with foreign characters in directory/file names (12 April 2008)
// 1.00 - released (24 March 2008)
//
// TERMS OF USE
// 
// This plugin is dual-licensed under the GNU General Public License and the MIT License and
// is copyright 2008 A Beautiful Site, LLC. 
//
if(jQuery) (function($){
	
	$.extend($.fn, {
		fileTree: function(o, h) {
			// Defaults
			if( !o ) var o = {};
			if( o.root == undefined ) o.root = '/';
			if( o.script == undefined ) o.script = 'jqueryFileTree.php';
			if( o.folderEvent == undefined ) o.folderEvent = 'click';
			if( o.expandSpeed == undefined ) o.expandSpeed= 250;
			if( o.collapseSpeed == undefined ) o.collapseSpeed= 250;
			if( o.expandEasing == undefined ) o.expandEasing = null;
			if( o.collapseEasing == undefined ) o.collapseEasing = null;
			if( o.multiFolder == undefined ) o.multiFolder = true;
			if( o.loadMessage == undefined ) o.loadMessage = 'Loading...';
			
			$(this).each( function() {
				
				function showTree(c, t) {
					$(c).addClass('wait');
					$(".jqueryFileTree.start").remove();
					$.post(o.script, { dir: t }, function(data) {
						$(c).find('.start').html('');
						$(c).removeClass('wait').append(data);
						if( o.root == t ) $(c).find('UL:hidden').show(); else $(c).find('UL:hidden').slideDown({ duration: o.expandSpeed, easing: o.expandEasing });
						bindTree(c);
					});
				}
				
				function bindTree(t) {
					$(t).find('LI A').bind(o.folderEvent, function() {
						if( $(this).parent().hasClass('directory') ) {
							if( $(this).parent().hasClass('collapsed') ) {
								// Expand
								if( !o.multiFolder ) {
									$(this).parent().parent().find('UL').slideUp({ duration: o.collapseSpeed, easing: o.collapseEasing });
									$(this).parent().parent().find('LI.directory').removeClass('expanded').addClass('collapsed');
								}
								$(this).parent().find('UL').remove(); // cleanup
								showTree( $(this).parent(), escape($(this).attr('rel').match( /.*\// )) );
								$(this).parent().removeClass('collapsed').addClass('expanded');
							} else {
								// Collapse
								$(this).parent().find('UL').slideUp({ duration: o.collapseSpeed, easing: o.collapseEasing });
								$(this).parent().removeClass('expanded').addClass('collapsed');
							}
						} else {
							h($(this).attr('rel'));
						}
						return false;
					});
					// Prevent A from triggering the # on non-click events
					if( o.folderEvent.toLowerCase != 'click' ) $(t).find('LI A').bind('click', function() { return false; });
				}
				// Loading message
				$(this).html('<ul class="jqueryFileTree start"><li class="wait">' + o.loadMessage + '<li></ul>');
				// Get the initial file list
				showTree( $(this), escape(o.root) );
			});
		}
	});
	
})(jQuery);

static/js/layout.js

(function($) {
    var views = {
        'topToolbar': {
            id: 'topToolbar',
            view: 'Box',
            background: 'theme(panel)',
            rect: '0 0 1000 59',
            anchors: 'top left right',
            childViews: [{
                id: 'btnNewProject',
                view: 'Button',
                rect: '700 16 100 24',
                anchors: 'top right',
                text: '+ New project'
            }, {
                id: 'btnVote',
                view: 'Box', src:'http://nodeknockout.com/images/voteko.png',
                anchors: 'top right',
                rect: '860 -3 126 63'
            },
            {
                view: 'Label', anchors: 'left top right', rect: '30 10 200 40', text: 'Nodejs in-browser development environment',
                style: {
                    fontSize: '30px',
                    color: '#000',
                    textShadow: '-1px -1px #dddddd'
                }
            }]
        }
    };

    function _panel(id, label, options) {
        options = options || {};
        //var rect =
        return {
            view: 'Box',
            rect: options.rect || '300 400',
            anchors: options.anchors || 'top left right bottom',
//            style: {border: '1px solid #00ff00'},
            childViews: [{   
                view: 'Box',
                background: 'theme(panel)',
                rect: '300 24',
                anchors: 'left top right',
                style: {
                        width: '100%'
                },
                childViews: [{
                    view: 'Label',
                    rect: '0 0 200 24',
                    anchors: 'left top right bottom',
                    text: label,
                    style: {
                        textAlign: 'center',
                        width: '100%',
                        fontWeight: 'bold'
                    }
                }]
            },
            {
                id: id,
                view: 'Box',
                rect: options.rect ? (function(rect) {
                    rect = rect.split(' ');
                    return [0, 25, rect[2], rect[3] - 25].join(' ');
                })(options.rect): '0 25 300 375',
                anchors: 'top left right bottom',
                style: {
//                    border: '1px dotted #ff0000',
                    width: '100%',
                    overflow: 'auto'
                }
            }
            ]
        };
    }

    function _editor(id, label, options) {
       options = options || {};
        //var rect =
        return {
            view: 'Box',
            rect: '0 0 700 500',
            anchors: 'top left right bottom',
//            style: {border: '1px solid #00ff00'},
            childViews: [{
                view: 'Box',
                background: 'theme(panel)',
                rect: '300 24',
                anchors: 'left top right',
                style: {
                        width: '100%'
                },
                childViews: [{
                    id: id+'Label',
                    view: 'Label',
                    rect: '0 0 200 24',
                    anchors: 'left top right bottom',
                    text: label,
                    style: {
                        textAlign: 'center',
                        width: '100%',
                        fontWeight: 'bold'
                    }
                }]
            },
            {
                id: id,
                view: 'Box',
                rect: '0 0 700 500',
                anchors: 'top left right bottom',
                style: {
                        width: '100%'
                }
            },
            {
                view:'Box',
//                background: 'theme(panel)',
                rect: '392 450 300 48',
                anchors: 'bottom right',
//                style: {border: '1px solid #00ff00'},
                childViews: [
                    {id:'btnSave', view: 'Button', rect: '110 22 60 24', anchors: 'left top', text: 'Save'},
                    {id:'btnRun', view: 'Button', rect: '180 22 100 24', anchors: 'left top', text: 'Run with node'},
                ]

            }
            ]
        };
    }

    // global namespace to hold all the objects

    $.ft = {
        buildLayout: function () {
            uki({   // create the main layout
                id: 'mainContainer',
                view: 'Box',
                rect: '1000 600',
                anchors: 'left top right bottom',
                minSize: '800 0',
                childViews: [
                views.topToolbar,
                {
                    view: 'HSplitPane',
                    rect: '0 60 1000 600',
                    anchors: 'left top right bottom',
                    handlePosition: 300,
                    leftMin: 200,
                    rightMin: 300,
                    // project browser and code navigator
                    leftChildViews: [{
                        view: 'VSplitPane',
                        rect: '300 600',
                        anchors: 'left top right bottom',
                        handlePosition: 400,
                        minSize: "200 300",
                        vertical: true,
                        topPane: _panel('panelProject', 'Project browser'),
                        bottomPane: _panel('panelNavigator', 'Code Navigator')
                    }],
                    // code editor and info window like console, monitor etc.
                    rightChildViews: [{
                        view: 'VSplitPane',
                        rect: '700 700',
                        anchors: 'left top right bottom',
//                        style: {border: '1px dashed blue'},
                        vertical: true,
                        handlePosition: 500,
                        minSize: "400 300",
                        topPane: _editor('codeEditor', 'Editor'),
                        bottomPane: _panel('panelConsole', 'Console',
                            {rect:'0 0 700 200'})
                    }]
                }]
            }).attachTo( window, '1000 600');
            // popup for buttons
            uki({id:'ppNewProject', view: 'Popup', rect: '0 20 240 120', anchors: 'right top', relativeTo: uki('#btnNewProject')[0]
                , childViews: [
                    {view: 'Label',    rect: '10 10 50 24', anchors: 'left top', html: 'Name:'},
                    {id:'inputProjectName', view: 'TextField',rect: '60 10 170 24', anchors: 'left top', value: "", placeholder: 'of your project'},
                    {view: 'Label',    rect: '10 50 50 24', anchors: 'left top', html: 'Url:'},
                    {id:'inputProjectUrl', view: 'TextField',rect: '60 50 170 24', anchors: 'left top', value: "", placeholder: 'of your GIT repository'},
                    {id:'btnGitClone', view: 'Button', rect: '160 86 70 24', anchors: 'right bottom', text: "Clone"},
                ]
            }).hide();
        },
        createEditorWrapper: function(id) {
           return uki({
               id: id,
               view: 'Box',
               rect: '0 25 700 470',
               anchors: 'top left right bottom',
                style: {
                    width: '100%',
                    overflow: 'auto'
                }
           });
        }
    }
})(jQuery);

